vulture_os/toolkit/api_parser/reachfive/reachfive.py

```diff
@@ -153,25 +176,35 @@ def execute(self):
             total = tmp_logs['total']
             logs = tmp_logs['items']
 
-            self.write_to_file([json.dumps(l) for l in tmp_logs['items']])
+            self.write_to_file([json.dumps(l) for l in logs])
             # Writting may take some while, so refresh token in Redis
             self.update_lock()
 
             page += 1
             cpt += len(logs)
 
             if len(logs) > 0:
+
                 # Get latest date returned from API (logs are often not properly ordered, even with a 'sort'ed query...)
                 assert logs[0]['date'], "logs don't have a 'date' field!"
                 assert logs[0]['date'][-1] == "Z", f"unexpected date format '{logs[0]['date']}', are they UTC?"
-                last_datetime = [last_datetime]
-                last_datetime.extend([x['date'] for x in logs])
+                last_datetime = [x['date'] for x in logs]
                 # ISO8601 timestamps are sortable as strings
                 last_datetime = max(last_datetime)
                 msg = f"most recent log is from {last_datetime}"
                 logger.debug(f"[{__parser__}]:execute: {msg}", extra={'frontend': str(self.frontend)})
 
                 # Replace "Z" by "+00:00" for datetime parsing
                 # No need to make_aware, date already contains timezone
-                self.frontend.last_api_call = datetime.fromisoformat(last_datetime.replace("Z", "+00:00"))
+                self.frontend.last_api_call = datetime.fromisoformat(last_datetime)
+
+            if cpt % 10000 == 0:
+                # Sleep 10 sec every 10 000 fetched logs to avoid too many requests
+                time.sleep(10)
+
+            if page == 11:  # Get max 10 000 logs to avoid API error
+                since = self.frontend.last_api_call.isoformat(timespec="milliseconds")
+                page = 1
+                cpt = 0
```

Comment:
seems like a complex answer to a simple problem : the collector should stop and continue querying the API at the next call
if more logs are generated than its possible to get from the API, the rate limits should be handled from the console, and configured accordingly

Reply:
This is the desired result